### ISSUE TITLE
Since MySQL's query optimisation is braindead, implement a way to force indeces

### ISSUE THREAD

**Issue by iglocska** [MISP/MISP#3321]
Something along the lines of this should work:

https://stackoverflow.com/questions/15949720/cakephp-force-index-at-model

**Comment by github-germ** [comment]
@iglocska Hi !! -- This is long posting. Hoping it's clear and that you can provide insight. Please bear with me...

Very interesting that I find this issue from nearly 4 years. Hoping you can provide some quick feedback on a performance problem we are experiencing with the `Search Attributes` WebUI page as I was starting to think a solution may be somehow having the `AttributeController` causing MySQL to use `FORCE (index}`. And then I stumbled onto this posting.

We see `Search Attributes` for a large list of IPs (1-4k) pasted in the expression textbox (`textarea name="data[Attribute][value]"`) sometimes returning instantly, other times taking 8 minutes (and sometimes much longer). Note that this a common use case for our Threat Intelligence Analysts.

I captured the `SELECT COUNT(*) ...` triggered in MySQL (I see two queries, the full attribute retrieval with `LIMIT 60` for the first WebUI page, and this count for the total pages). When the results are taking many minutes, `EXPLAIN SELECT COUNT(*)` shows the `events` table the first in the `JOIN` with the key being `attributes.event_id` for the `attributes` table; however, when the results are fast, `attributes` is first with `(value1,value2)` as the keys.

When we add `FORCE(value1,value2)` to the `SELECT` and run manually while the db result is slow, the result is fast (as we'd expect).

So before seeing your issue here, my hypothesis was similar to yours above. I was thinking that the MySQL query optimization logic was biased to `event` vs. `attribute` as most transactions were occuring with an `event` focus, and when this `Search Attributes` occured that result was penalized since the faster keys were not currently considered optimal.

I even ran `optimize table attributes` in MySQL and this caused `attribute` to be the key bias and results were fast, but given more time with many other transactions eventually the db engine set key bias back to `event`. This leads me to believe that without some forcing of keys some transactions will be fast and others slow without any method to improve performance.

BOTTOM LINE:
* Are my findings and hypothesis close to correct?
* If not, please help me better understand?
* If yes, is there a solution in PHP or Redis to help, and if so, is there any chance for a "quick fix"?

Many thanks!!

**Comment by iglocska** [comment]
It sounds correct indeed that was my guess too. We used to have certain forced indeces in some of the critical places that we identified (or actually forcing an index is tricky, however there was a hacky way around the limitation). 

However, most of the really bad query optimisations seemed to have come from older mysql / mariadb versions and we've dropped it. 

It's a good find though, I think we can safely say that if value fields are set as lookup values we should always prioritise them!

**Comment by github-germ** [comment]
@iglocska -- Thanks for confirming my findings with yours. Are there plans to force the index keys for attribute focused searches in a release soon?

**Comment by righel** [comment]
Hello @github-germ 
I was able to reproduce the issue and I'm working on a fix, I will keep you posted. 
Sorry for the delay.

**Comment by github-germ** [comment]
@righel -- Thanks!!

**Comment by righel** [comment]
@github-germ I merged this PR: https://github.com/MISP/MISP/pull/8390 to the `develop` branch.
Could you please switch to `develop` branch and test if this fixes your issue?
You also need to update the database config file to use the patched MySQL datasource. Change `Database/Mysql` to `Database/MysqlExtended`.

The config file should look like this:
```php
<?php
  class DATABASE_CONFIG {
          public $default = array(
                  'datasource' => 'Database/MysqlExtended',
                  ...
```

**Comment by github-germ** [comment]
@righel -- Excellent... We are working on a internal release hand-off so I will be delayed a bit. I hope to build our docker image from `develop` along with the `database.php` mod by Monday/Tuesday and report back. Many thanks!! Stay tuned...

**Comment by github-germ** [comment]
@righel -- I've loaded the `develop` branch. Here's what the mysql `expect select ...` shows for the same attribute query we have for this test case. Does this match your expectation on what this fix should produce? Notice the `Using sort_union(value1,value2);` in the `Extra` column.
```
+----+-------------+--------------+------------+-------------+-----------------------------------------------------------+---------------+---------+---------------------------------+--------+----------+----------------------------------------------+
| id | select_type | table        | partitions | type        | possible_keys                                             | key           | key_len | ref                             | rows   | filtered | Extra                                        |
+----+-------------+--------------+------------+-------------+-----------------------------------------------------------+---------------+---------+---------------------------------+--------+----------+----------------------------------------------+
|  1 | SIMPLE      | Attribute    | NULL       | index_merge | event_id,sharing_group_id,value1,value2,object_id,deleted | value1,value2 | 767,767 | NULL                            | 746000 |    50.00 | Using sort_union(value1,value2); Using where |
|  1 | SIMPLE      | Event        | NULL       | eq_ref      | PRIMARY,sharing_group_id,org_id                           | PRIMARY       | 4       | misp.Attribute.event_id         |      1 |    20.95 | Using where                                  |
|  1 | SIMPLE      | SharingGroup | NULL       | eq_ref      | PRIMARY                                                   | PRIMARY       | 4       | misp.Attribute.sharing_group_id |      1 |   100.00 | Using index                                  |
|  1 | SIMPLE      | Object       | NULL       | eq_ref      | PRIMARY                                                   | PRIMARY       | 4       | misp.Attribute.object_id        |      1 |   100.00 | Using where                                  |
+----+-------------+--------------+------------+-------------+-----------------------------------------------------------+---------------+---------+---------------------------------+--------+----------+----------------------------------------------+
```

**Comment by github-germ** [comment]
Btw, I hadn't thought that the db schema was update from 85 in 2.4.158 and 86 in develop. And now I cannot move my instance back to our 2.4.158 container as the db and code are incompatible. This becomes my lesson learned which leads to the question on when you think 2.4.159 including this fix (if it is good) might be forthcoming?

THANKS!

**Comment by righel** [comment]
> @righel -- I've loaded the `develop` branch. Here's what the mysql `expect select ...` shows for the same attribute query we have for this test case. Does this match your expectation on what this fix should produce? ```

Yes, with the hinted query index the results should be consistent now. If you search attributes and value filters are present it should use always that index. Do you still get long running requests?

Regarding the db update, the only change introduced in db version 86 is a index on the attributes timestamp column. What do you mean by the db and code are incompatible? are you getting an error?

**Comment by github-germ** [comment]
The long running requests are performing well. However, only after the MySQL heuristics change the index bias for attribute selects are changed does performance degrade significantly -- this "flip" occurs seemingly randomly sometimes quickly, other times after a day or so, all without notice. Now that the index hint exists I assume this will not occur and attribute searches will perform well.

Good to know the timestamp index fix is included. Will check out that performance gain which may be a hge benefit some of our apps. THANKS.

Re' db... there were lots of errors from our 2.4.158 container entrypoint scripts launched. I believe these were occuring in our PyMISP apps (PyMISP was still built with 2.4.157). I needed to get the instance running so I re-deployed the `develop` container, and it is running. If I have time, I will revisit those errors and report back.

**Comment by github-germ** [comment]
I know you don't generally advertise predictions for the date of the next official release. For our planning purposes does your crystal ball seem to hint to a release within 2 weeks for these 2 fixes? :-)

**Comment by righel** [comment]
I'm pretty confident there will be a release early next week.

**Comment by righel** [comment]
Hi @github-germ this fix and many others are in the latest release `2.4.159`
